pymc/tests/test_distributions_moments.py

```diff
@@ -62,13 +62,70 @@
     ZeroInflatedNegativeBinomial,
     ZeroInflatedPoisson,
 )
-from pymc.distributions.distribution import get_moment
+from pymc.distributions.distribution import _get_moment, get_moment
+from pymc.distributions.logprob import logpt
 from pymc.distributions.multivariate import MvNormal
 from pymc.distributions.shape_utils import rv_size_is_none, to_tuple
 from pymc.initial_point import make_initial_point_fn
 from pymc.model import Model
 
 
+def test_all_distributions_have_moments():
+    import pymc.distributions as dist_module
+
+    from pymc.distributions.distribution import DistributionMeta
+
+    dists = (getattr(dist_module, dist) for dist in dist_module.__all__)
+    dists = (dist for dist in dists if isinstance(dist, DistributionMeta))
+    missing_moments = {
+        dist for dist in dists if type(getattr(dist, "rv_op", None)) not in _get_moment.registry
+    }
+
+    # Ignore super classes
+    missing_moments -= {
+        dist_module.Distribution,
+        dist_module.Discrete,
+        dist_module.Continuous,
+        dist_module.NoDistribution,
+        dist_module.DensityDist,
+        dist_module.simulator.Simulator,
+    }
+
+    # Distributions that have not been refactored for V4 yet
+    not_implemented = {
+        dist_module.multivariate.LKJCorr,
+        dist_module.mixture.Mixture,
+        dist_module.mixture.MixtureSameFamily,
+        dist_module.mixture.NormalMixture,
+        dist_module.timeseries.AR,
+        dist_module.timeseries.AR1,
+        dist_module.timeseries.GARCH11,
+        dist_module.timeseries.GaussianRandomWalk,
+        dist_module.timeseries.MvGaussianRandomWalk,
+        dist_module.timeseries.MvStudentTRandomWalk,
+    }
+
+    # Distributions that have been refactored but don't yet have moments
+    not_implemented |= {
+        dist_module.discrete.DiscreteWeibull,
+        dist_module.multivariate.CAR,
+        dist_module.multivariate.DirichletMultinomial,
+        dist_module.multivariate.KroneckerNormal,
+        dist_module.multivariate.Wishart,
+    }
+
+    # Assert that they do not have moments yet
+    assert missing_moments.intersection(not_implemented) == not_implemented
+
+    missing_moments -= not_implemented
+
+    if missing_moments:
+        raise NotImplementedError(
+            "The following distributions do not have a moment method:\n"
+            + "\n".join(sorted(map(str, list(missing_moments))))
+        )
```

Comment:
can we make the set operations a little easier to comprehend?

```suggestion
    unexpected_implemented = not_implemented - missing_moments
    unexpected_not_implemented = missing_moments - not_implemented
    
    if unexpected_implemented:
        raise Exception(
            f"Distributions {unexpected_implemented}" have a `get_moment` implemented."
            " That's good! But this test must be updated."
        )
        
    if unexpected_not_implemented:
        raise NotImplementedError(
            "Unexpected by this test, some distributions do not have a `get_moment` implementation:\n"
            + "\n".join(sorted(map(str, list(unexpected_not_implemented))))
        )
```

I didn't test this and my syntax probably doesn't pass the pre-commit. So please apply locally..

Reply:
Updated!